data-analytics/dataflow-python-examples/tests/test_data_generator.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#            http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import logging
+import datetime
+import unittest
+import os
+import re
+
+from faker_schema.faker_schema import FakerSchema
+
+from dataflow_python_examples.data_generation_for_benchmarking import DataGenerator, FakeRowGen
+
+
+class TestDataGenerator(unittest.TestCase):
+    """The test cases are focused on the business logic.  In this case this is how we parse the
+    schemas, generate data and label images.
+
+    Execution Note:
+        This script is stored in professional-services/data-analytics/dataflow-python-examples/tests
+        but should be copied to professional-services/data-analytics/dataflow-python-examples/ and
+        run from there.
+    """
+
+    def setUp(self):
+        # User parser to define default data_args.
+
+        # Note changed default for schema_file for ease of testing.
+        dir_path = os.path.dirname(os.path.realpath(''))
+        schema_file = os.path.join(dir_path, 'dataflow-python-examples', 'dataflow_python_examples',
+                                   'resources', 'lineorder-schema.json')
+
+        self.data_gen = DataGenerator(bq_schema_filename=schema_file,
+                                      p_null=0.0, n_keys=1000, min_date='2000-01-01',
+                                      max_date=datetime.date.today().strftime('%Y-%m-%d'),
+                                      only_pos=True, max_int=10**11, max_float=float(10**11),
+                                      float_precision=2, write_disp='WRITE_APPEND')
+
+
+        self.fakerowgen = FakeRowGen(self.data_gen)
+
+
+        logging.basicConfig(level=logging.DEBUG)
+
+    def test_get_bq_schema_string(self):
+        """
+        This tests the get_bq_schema_string method of the DataGenerator class which parses
+        a 'fieldname:field_type' string defining a the schema from a
+        [{name,type,mode}] schema dictionary.
+        """
+        expected_bq_schema_string = "lo_order_key:STRING,lo_linenumber:INTEGER,lo_part_key:STRING,lo_cust_key:STRING,lo_orderdate:DATE,lo_revenue:FLOAT,lo_supp_key:STRING,lo_quantity:INTEGER,lo_extendedprice:FLOAT,lo_discount:FLOAT,lo_supplycost:FLOAT,lo_ordpriority:INTEGER,lo_ordtotalprice:FLOAT,lo_shippriority:INTEGER,lo_tax:FLOAT,lo_shipmode:FLOAT,lo_recieptfile:STRING"
```

Comment:
Exceeds the 80 column limit for a string. Either break it up with implicit string concatenation, or a string join

```
expected_bq_schema_string = ','.join([
  'lo_order_key:STRING',
  'lo_linenumber:INTEGER',
  ...
])
```

Reply:
Changing to suggested string join for clarity.